core/sds-aspect-model-validator/src/test/java/io/openmanufacturing/sds/aspectmodel/shacl/ShaclValidatorTest.java

```diff
@@ -0,0 +1,1737 @@
+      final Shape.Property property = shape.properties().get( 0 );
+      assertThat( property.attributes().name() ).hasValue( "Test property" );
+      assertThat( property.attributes().description() ).hasValue( "Test description" );
+      assertThat( property.attributes().constraints() ).hasOnlyElementsOfTypes( DatatypeConstraint.class, MinCountConstraint.class );
+      assertThat( property.attributes().constraints() ).hasAtLeastOneElementOfType( DatatypeConstraint.class );
+      assertThat( property.attributes().constraints() ).hasAtLeastOneElementOfType( MinCountConstraint.class );
+
+      assertThat( property.path() ).isEqualTo( new PredicatePath( ResourceFactory.createProperty( namespace + "testProperty" ) ) );
+   }
+
+   @Test
+   public void testClassConstraintEvaluation() {
+      final Model shapesModel = model( """
+            @prefix sh: <http://www.w3.org/ns/shacl#> .
+            @prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
+            @prefix : <http://example.com#> .
+
+            :MyShape
+               a sh:NodeShape ;
+               sh:targetClass :TestClass ;
+               sh:name "Test shape" ;
+               sh:description "Test shape description" ;
+               sh:property [
+                  sh:path :testProperty ;
+                  sh:class :TestClass2 ;
+               ] .
+            """ );
+
+      final Model dataModel = model( """
+            @prefix : <http://example.com#> .
+            @prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
+            @prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
+
+            :MySuperType a :TestClass2 .
+            :MyType rdfs:subClassOf :MySuperType .
+
+            :Bar a :TestClass ;
+              :testProperty [ a :MyType ] .
+
+            :Foo a :TestClass ;
+              :testProperty [ a :SomethingElse ] .
+            """ );
+
+      final ShaclValidator validator = new ShaclValidator( shapesModel );
+      final Resource element = dataModel.createResource( namespace + "Foo" );
+      final List<Violation> violations = validator.validateElement( element );
+
+      assertThat( violations.size() ).isEqualTo( 1 );
+      final Violation finding = violations.get( 0 );
+      assertThat( finding ).isInstanceOf( ClassTypeViolation.class );
+      final ClassTypeViolation violation = (ClassTypeViolation) finding;
+      assertThat( violation.context().element() ).isEqualTo( element );
+      assertThat( violation.context().shape().attributes().uri() ).hasValue( namespace + "MyShape" );
+      assertThat( violation.propertyName() ).isEqualTo( ":testProperty" );
+      assertThat( violation.elementName() ).isEqualTo( ":Foo" );
+      assertThat( violation.actualClass().getURI() ).isEqualTo( "http://example.com#SomethingElse" );
+      assertThat( violation.allowedClass().getURI() ).isEqualTo( "http://example.com#TestClass2" );
+      assertThat( violation.message() ).isNotEmpty();
```

Comment:
Any reason you do not check the correctness of the message?

Reply:
The reason was not to break tests when the messages are changed, i.e., a violation would not be invalid just because it's message text changes. But since this also tests the message interpolation, it's probably good idea to do it.